test/e2e/aks_adopt.go

```diff
@@ -0,0 +1,154 @@
+			g.Expect(obj.GetAnnotations()).NotTo(HaveKey(clusterctlv1.BlockMoveAnnotation))
+		}, waitForBlockMoveGone...).Should(Succeed())
+	}
+
+	removeFinalizers := func(obj client.Object) {
+		Eventually(func(g Gomega) {
+			err := mgmtClient.Get(ctx, client.ObjectKeyFromObject(obj), obj)
+			g.Expect(err).NotTo(HaveOccurred())
+			obj.SetFinalizers([]string{})
+			err = mgmtClient.Update(ctx, obj)
+			g.Expect(err).NotTo(HaveOccurred())
+		}, updateResource...).Should(Succeed())
+	}
+
+	waitForImmediateDelete := []any{"30s", "5s"}
+	beginDelete := func(obj client.Object) {
+		Eventually(func(g Gomega) {
+			err := mgmtClient.Delete(ctx, obj)
+			g.Expect(err).NotTo(HaveOccurred())
+		}, updateResource...).Should(Succeed())
+	}
+	shouldNotExist := func(obj client.Object) {
+		waitForGone := []any{"30s", "5s"}
+		Eventually(func(g Gomega) {
+			err := mgmtClient.Get(ctx, client.ObjectKeyFromObject(obj), obj)
+			g.Expect(apierrors.IsNotFound(err)).To(BeTrue())
+		}, waitForGone...).Should(Succeed())
+	}
+	deleteAndWait := func(obj client.Object) {
+		Eventually(func(g Gomega) {
+			err := mgmtClient.Delete(ctx, obj)
+			g.Expect(apierrors.IsNotFound(err)).To(BeTrue())
+		}, waitForImmediateDelete...).Should(Succeed())
+	}
+
+	cluster := input.Cluster
+	Eventually(func(g Gomega) {
+		err := mgmtClient.Get(ctx, client.ObjectKeyFromObject(cluster), cluster)
+		g.Expect(err).NotTo(HaveOccurred())
+		cluster.Spec.Paused = true
+		err = mgmtClient.Update(ctx, cluster)
+		g.Expect(err).NotTo(HaveOccurred())
+	}, updateResource...).Should(Succeed())
+
+	// wait for the pause to take effect before deleting anything
+	amcp := &infrav1.AzureManagedControlPlane{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cluster.Spec.ControlPlaneRef.Namespace,
+			Name:      cluster.Spec.ControlPlaneRef.Name,
+		},
+	}
+	waitForNoBlockMove(amcp)
+	for _, mp := range input.MachinePools {
+		ammp := &infrav1.AzureManagedMachinePool{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: mp.Spec.Template.Spec.InfrastructureRef.Namespace,
+				Name:      mp.Spec.Template.Spec.InfrastructureRef.Name,
+			},
+		}
+		waitForNoBlockMove(ammp)
+	}
+
+	beginDelete(cluster)
+
+	for _, mp := range input.MachinePools {
+		beginDelete(mp)
+
+		ammp := &infrav1.AzureManagedMachinePool{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: mp.Spec.Template.Spec.InfrastructureRef.Namespace,
+				Name:      mp.Spec.Template.Spec.InfrastructureRef.Name,
+			},
+		}
+		removeFinalizers(ammp)
+		deleteAndWait(ammp)
+
+		removeFinalizers(mp)
+		shouldNotExist(mp)
+	}
+
+	removeFinalizers(amcp)
+	deleteAndWait(amcp)
+	// AzureManagedCluster never gets a finalizer
+	deleteAndWait(&infrav1.AzureManagedCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: cluster.Spec.InfrastructureRef.Namespace,
+			Name:      cluster.Spec.InfrastructureRef.Name,
+		},
+	})
+
+	removeFinalizers(cluster)
+	shouldNotExist(cluster)
+
+	clusterctl.ApplyClusterTemplateAndWait(ctx, input.ApplyInput, input.ApplyResult)
```

Comment:
I'm a bit confused on how this test is laid out. Do we orphan this cluster and then adopt it back as a CAPZ cluster? I'm a bit confused where the adoption occurs since the logic here seems to be only related to orphaning the cluster.

Reply:
It could help to add a code comment to separate the logic between orphaning/adopting the cluster.

Reply:
> Do we orphan this cluster and then adopt it back as a CAPZ cluster?

Yes, that's how I tried to describe it in the doc comment for this function.

> I'm a bit confused where the adoption occurs since the logic here seems to be only related to orphaning the cluster.

The ApplyAndWait at the very end is the adoption. "Adoption" isn't really its own separate thing, it's only applying templates that happen to match existing resources.